### PR TITLE
8341052: SHA-512 implementation using SHA-NI

### PR DESCRIPTION
**Pull request by smita-kamath**
Hi, I want to submit an optimization for SHA-512 algorithm using SHA instructions (sha512msg1, sha512msg2 and sha512rnds2) . Kindly review the code and provide feedback. Thank you.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8341052](https://bugs.openjdk.org/browse/JDK-8341052): SHA-512 implementation using SHA-NI (**Enhancement** - P4)

### Reviewers
 * [Sandhya Viswanathan](https://openjdk.org/census#sviswanathan) (@sviswa7 - **Reviewer**) Review applies to [85c1aea9](https://git.openjdk.org/jdk/pull/20633/files/85c1aea90b10014aa34dfc902dff2bfd31bd70c0)
 * [Srinivas Vamsi Parasa](https://openjdk.org/census#sparasa) (@vamsi-parasa - Author) Review applies to [85c1aea9](https://git.openjdk.org/jdk/pull/20633/files/85c1aea90b10014aa34dfc902dff2bfd31bd70c0)
 * [Jatin Bhateja](https://openjdk.org/census#jbhateja) (@jatin-bhateja - **Reviewer**)
 * [Anthony Scarpino](https://openjdk.org/census#ascarpino) (@ascarpino - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/20633/head:pull/20633` \
`$ git checkout pull/20633`

Update a local copy of the PR: \
`$ git checkout pull/20633` \
`$ git pull https://git.openjdk.org/jdk.git pull/20633/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 20633`

View PR using the GUI difftool: \
`$ git pr show -t 20633`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/20633.diff">https://git.openjdk.org/jdk/pull/20633.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/20633#issuecomment-2383546400)

### CODE REVIEW

File: src/hotspot/cpu/x86/stubGenerator_x86_64.cpp [hunk 1]
```diff
@@ -1556,7 +1556,7 @@ address StubGenerator::generate_sha256_implCompress(bool multi_block, const char
 }
 
 address StubGenerator::generate_sha512_implCompress(bool multi_block, const char *name) {
-  assert(VM_Version::supports_avx2(), "");
+  assert(VM_Version::supports_avx2() || VM_Version::supports_sha512(), "");
   assert(VM_Version::supports_bmi2(), "");
```
Comment (sviswa7) [inline src/hotspot/cpu/x86/stubGenerator_x86_64.cpp:1560]: The SHA512 algorithm is also using AVX2 instructions but doesn't need bmi2.
So this needs to be changed to:
 assert(VM_Version::supports_avx2(), "");
 assert(VM_Version::supports_bmi2()  || VM_Version::supports_sha512(), "");

**Comment by sviswa7** [comment]
The line 1306 in vm_version_x86.cpp needs to be changed from:
 if (UseSHA && supports_avx2() && supports_bmi2()) {
to
  if (UseSHA && supports_avx2() && (supports_bmi2() || supports_sha512())) {

File: src/hotspot/cpu/x86/macroAssembler_x86_sha.cpp [hunk 2]
```diff
@@ -1519,5 +1519,183 @@ void MacroAssembler::sha512_AVX2(XMMRegister msg, XMMRegister state0, XMMRegiste
     }
 }
 
+void MacroAssembler::sha512_update_ni_x1(Register arg_hash, Register arg_msg, Register ofs, Register limit, bool multi_block) {
+    Label done_hash, block_loop;
+    address K512_W = StubRoutines::x86::k512_W_addr();
+
+    vbroadcasti128(xmm15, ExternalAddress(StubRoutines::x86::pshuffle_byte_flip_mask_addr_sha512()), rbx);
```
Comment (sviswa7) [inline src/hotspot/cpu/x86/macroAssembler_x86_sha.cpp:1526]: rbx is save on entry register. Better to use a save on call register as temp, say r10.

File: src/hotspot/cpu/x86/assembler_x86.cpp [hunk 3]
```diff
@@ -6762,6 +6762,27 @@ void Assembler::sha256msg2(XMMRegister dst, XMMRegister src) {
   emit_int16((unsigned char)0xCD, (0xC0 | encode));
 }
 
+void Assembler::sha512msg1(XMMRegister dst, XMMRegister src) {
+  assert(VM_Version::supports_sha512(), "");
```
Comment (sviswa7) [inline src/hotspot/cpu/x86/assembler_x86.cpp:6766]: For all the new sha512 instructions, this should be:
assert(VM_Version::supports_sha512() && VM_Version::supports_avx(), "");

File: src/hotspot/cpu/x86/assembler_x86.cpp [hunk 4]
```diff
@@ -6762,6 +6762,27 @@ void Assembler::sha256msg2(XMMRegister dst, XMMRegister src) {
...
+  InstructionAttr attributes(AVX_256bit, /* vex_w */ false, /* legacy_mode */ false, /* no_mask_reg */ true, /* uses_vl */ false);
```
Comment (sviswa7) [inline src/hotspot/cpu/x86/assembler_x86.cpp:6767]: For all the new sha512 instructions legacy_mode should be true.

File: src/hotspot/cpu/x86/assembler_x86.cpp [hunk 5]
```diff
@@ -11672,6 +11693,17 @@ void Assembler::evbroadcasti64x2(XMMRegister dst, Address src, int vector_len) {
   emit_operand(dst, src, 0);
 }
 
+void Assembler::vbroadcasti128(XMMRegister dst, Address src) {
```
Comment (sviswa7) [inline src/hotspot/cpu/x86/assembler_x86.cpp:11696]: Please see vbroadcastf128, good to take vector_len as argument here also and have an assert to confirm that it is AVX-256 bit. 
Also need:
attributes.set_address_attributes(/* tuple_type */ EVEX_T4, /* input_size_in_bits */ EVEX_32bit);

**Comment by sviswa7** [comment]
The following comment is still pending:
The line 1306 in vm_version_x86.cpp needs to be changed from:
if (UseSHA && supports_avx2() && supports_bmi2()) {
to
if (UseSHA && supports_avx2() && (supports_bmi2() || supports_sha512())) {

**Review by sviswa7 — APPROVED** [review]

**Comment by sviswa7** [comment]
Looks good to me.

File: src/hotspot/cpu/x86/macroAssembler_x86_sha.cpp [hunk 6]
```diff
@@ -1519,5 +1519,183 @@ void MacroAssembler::sha512_AVX2(XMMRegister msg, XMMRegister state0, XMMRegiste
     }
 }
 
+void MacroAssembler::sha512_update_ni_x1(Register arg_hash, Register arg_msg, Register ofs, Register limit, bool multi_block) {
```
Comment (jatin-bhateja) [inline src/hotspot/cpu/x86/macroAssembler_x86_sha.cpp:1522]: Please add a comment on this mentioning the source of algorithm.

File: src/hotspot/cpu/x86/macroAssembler_x86_sha.cpp [hunk 7]
```diff
@@ -1519,5 +1519,183 @@ void MacroAssembler::sha512_AVX2(XMMRegister msg, XMMRegister state0, XMMRegiste
...
+      sha512rnds2(xmm12, xmm11, xmm0);
+      vperm2i128(xmm0, xmm0, xmm0, 0x01);
+      sha512rnds2(xmm11, xmm12, xmm0);
+      sha512msg1(xmm6, xmm3); //ymm6 = W[12..15] + S0(W[13..16])
+      i += 1;
+      //R20 - R23, R36 - R39, R52 - R55
+      vpaddq(xmm0, xmm4, Address(rax, i * 32), Assembler::AVX_256bit);
+      vpermq(xmm8, xmm4, 0x1b, Assembler::AVX_256bit);//ymm8 = W[20] W[21] W[22] W[23]
+      vpermq(xmm9, xmm3, 0x39, Assembler::AVX_256bit);//ymm9 = W[16] W[19] W[18] W[17]
+      vpblendd(xmm7, xmm8, xmm9, 0x3f, Assembler::AVX_256bit);//ymm7 = W[20] W[19] W[18] W[17]
```
[the remark this answers is not in the document]
Comment (smita-kamath) [reply src/hotspot/cpu/x86/macroAssembler_x86_sha.cpp:1603]: We can do this change in a separate PR.

**Review by vamsi-parasa — APPROVED** [review]
This implementation looks good to me. 

I went through the implementation of `sha512_update_ni_x1`. Looked at it line by line and compared it to the ipsec [implementation](https://github.com/intel/intel-ipsec-mb/blob/main/lib/avx2_t4/sha512_x1_ni_avx2.asm). 

Thanks,
Srinivas Vamsi Parasa (Intel)

File: src/hotspot/cpu/x86/macroAssembler_x86_sha.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (vamsi-parasa) [reply src/hotspot/cpu/x86/macroAssembler_x86_sha.cpp:1603]: I agree with Smita. The current implementation has a one-to-one correspondence with the ipsec implementation. Any new changes or refactoring will require a new round of exhaustive testing and could be implemented as a separate PR.

[the remark this answers is not in the document]
Comment (jatin-bhateja) [reply src/hotspot/cpu/x86/macroAssembler_x86_sha.cpp:1603]: I agree, in principle, any optimization crafted to AVX2 is also applicable to AVX512 target, in future with AVX10.2 (converged ISA) we will have a 256bits flavors of two table permute for non-AVX512 targets, for now AVX-SHA512 is only available on client parts (upcoming Lunar lake) and its ok to follow the IPsec algorithm in toto.

**Review by jatin-bhateja — APPROVED** [review]

**Comment by smita-kamath** [comment]
@ascarpino, I have approvals for this PR. Would it be possible for you to run tests and let me know the results? I appreciate your help. Thanks.

**Review by ascarpino — APPROVED** [review]
Tier 1-3 passed on windows-x64, linux-x64, and macos-aarch64

**Pull request closed without merging**